### PR TITLE
Support for universal and ball joint in sdf

### PR DESCRIPTION
**Pull request by quarkytale**
Mapped universal and ball joint in sdf to be floating joints in urdf.

Questions:
1. Should the base be galactic or ros2?
2. Initially I added comments beside the change that `// Will need transforms published to function correctly`, but that's true for all joints in that list. Not sure if I should indicate that here since there'll be an example world (WIP) to demo the usage.

### CODE REVIEW

**Comment by sloretz** [comment]
> Should the base be galactic or ros2?

Please target the `ros2` branch. ROS projects always do development on the latest branch, and then backport as needed.

File: sdformat_urdf/src/sdformat_urdf.cpp
Comment (sloretz) [inline sdformat_urdf/src/sdformat_urdf.cpp:499]: I'd recommend making all joints types in SDFormat that don't have URDF equivalents map to `urdf::Joint::FLOATING` (except `sdf::JointType::INVALID`). That way other models with those joints could still be used as long as someone was willing to write a custom TF publisher for them.

File: sdformat_urdf/src/sdformat_urdf.cpp
Comment (sloretz) [inline sdformat_urdf/src/sdformat_urdf.cpp]: Nit, I recommend using `switch/case` fall-through to de-duplicate the code assigning `urdf_joint_->type`.

```suggestion
    case sdf::JointType::UNIVERSAL:  // Fall through
    case sdf::JointType::BALL:
      urdf_joint->type = urdf::Joint::FLOATING;
      break;
```

**Review by sloretz — CHANGES_REQUESTED** [review]
The logic looks good to me. Mind updating the tests so CI passes with the new behavior?

**Comment by quarkytale** [comment]
Tests should be passing now! The gearbox joint has a kinematic loop, which is expected. (I think a custom TF publisher will handle that or should I just make it invalid?)

File: sdformat_urdf/src/sdformat_urdf.cpp
Comment (chapulina) [inline sdformat_urdf/src/sdformat_urdf.cpp:514]: I think it may be worth it adding a check here to make sure `nullptr != sdf_axis` just to be safe and we don't risk accessing a null pointer below.

File: sdformat_urdf/test/joint_tests.cpp
Comment (chapulina) [inline sdformat_urdf/test/joint_tests.cpp:44]: I think lines 40~44 can be `EXPECT_EQ`, because we're not trying to access those pointers in the test anyway. This way, if in the future one of them fails, we will know whether the others fail too.

**Review by chapulina — APPROVED** [review]
I just have a couple of minor suggestions, but this looks good to me otherwise. I'd wait for an approval from @sloretz before merging.

After it's merged, we'll need to open a backport to `galactic`.

File: sdformat_urdf/src/sdformat_urdf.cpp
Comment (sloretz) [reply sdformat_urdf/src/sdformat_urdf.cpp:514]: This seems like a good idea.

File: sdformat_urdf/test/joint_tests.cpp
Comment (sloretz) [reply sdformat_urdf/test/joint_tests.cpp:44]: I agree, though this could already be improved about the existing tests in this file so I wouldn't block on it.

**Review by sloretz — APPROVED** [review]
LGTM. @chapulina has good points, but since they're existing issues I won't block on it.

I don't think we need a run of ci.ros2.org since this is outside of the usual repos file and doesn't have a quality level declaration. I also have no idea what would happen if we tried to build this on Windows.

File: sdformat_urdf/src/sdformat_urdf.cpp
Comment (quarkytale) [reply sdformat_urdf/src/sdformat_urdf.cpp:514]: Added the check and a test with a new sdf whose joint has no axis tag

File: sdformat_urdf/test/joint_tests.cpp
Comment (quarkytale) [reply sdformat_urdf/test/joint_tests.cpp:44]: I can work on this in a separate PR

Comment (quarkytale) [reply sdformat_urdf/test/joint_tests.cpp:44]: I can work on this in a separate PR :)

File: sdformat_urdf/src/sdformat_urdf.cpp
Comment (chapulina) [reply sdformat_urdf/src/sdformat_urdf.cpp:514]: Thanks for the new test and the check.

I'm not sure about treating this as an error though. I was looking at the SDF spec and it looks like the `axis` element is [optional](http://sdformat.org/spec?ver=1.9&elem=joint#joint_axis). That's probably because it depends on the joint type. I can see an argument for considering a missing axis an error for joint types that do have axes, but I think that should be handled at the `libsdformat` level, not here.

Is it possible to not return an `nullptr` here, and instead just skip the axis parsing if it's not available?

Comment (quarkytale) [reply sdformat_urdf/src/sdformat_urdf.cpp:514]: Wasn't aware of that, thanks for catching this! But by filtering out FLOATING and FIXED urdf joints, the sdf mapped counterparts reach that point are only continuous, revolute and prismatic. Those should be having an axis and the parser should fail without it?

Comment (chapulina) [reply sdformat_urdf/src/sdformat_urdf.cpp:514]: Ok, let's go with the current implementation, I think it's the SDF spec's fault for not being clear about what to do in this case. I'll see if I open an issue there later, but I don't want to block this on that.

**Review by chapulina — APPROVED** [review]

**Comment by quarkytale** [comment]
@mergifyio backport galactic

File: sdformat_urdf/test/joint_tests.cpp
Comment (quarkytale) [reply sdformat_urdf/test/joint_tests.cpp:44]: Opened #16 for improving tests.

**Pull request merged**
